tests/tests/classes/PluginManager.js

```diff
@@ -0,0 +1,281 @@
+      this.deployedResources = 0;
+    }
+
+    functions() {
+      return new Promise((resolve, reject) => {
+        this.deployedFunctions = this.deployedFunctions + 1;
+        return resolve();
+      });
+    }
+
+    resources() {
+      return new Promise((resolve, reject) => {
+        this.deployedResources = this.deployedResources + 1;
+        return resolve();
+      });
+    }
+  }
+
+  class SynchronousPluginMock {
+    constructor() {
+      this.commands = {
+        deploy: {
+          usage: 'Deploy to the default infrastructure',
+          lifeCycleEvents: [
+            'resources',
+            'functions'
+          ],
+          commands: {
+            onpremises: {
+              usage: 'Deploy to your On-Premises infrastructure',
+              lifeCycleEvents: [
+                'resources',
+                'functions'
+              ],
+            },
+          },
+        },
+      };
+
+      this.hooks = {
+        'deploy:functions': this.functions.bind(this),
+        'before:deploy:onpremises:functions': this.resources.bind(this),
+      };
+
+      // used to test if the function was executed correctly
+      this.deployedFunctions = 0;
+      this.deployedResources = 0;
+    }
+
+    functions() {
+      this.deployedFunctions = this.deployedFunctions + 1;
+    }
+
+    resources() {
+      this.deployedResources = this.deployedResources + 1;
+    }
+  }
+
+  beforeEach(() => {
+    serverless = new Serverless({});
+    pluginManager = new PluginManager(serverless);
+    helloWorld = new HelloWorld();
+  });
+
+  describe('#constructor()', () => {
+    it('should set the serverless instance', () => {
+      expect(pluginManager.serverless).to.deep.equal(serverless);
+    });
+
+    it('should create an empty plugins array', () => {
+      expect(pluginManager.plugins.length).to.equal(0);
+    });
+
+    it('should create an empty commandsList array', () => {
+      expect(pluginManager.commandsList.length).to.equal(0);
+    });
+
+    it('should create an empty commands object', () => {
+      expect(pluginManager.commands).to.deep.equal({});
+    });
+  });
+
+  describe('#addPlugin()', () => {
+    it('should add a plugin instance to the plugins array', () => {
+      pluginManager.addPlugin(SynchronousPluginMock);
+
+      expect(pluginManager.plugins[0]).to.be.an.instanceof(SynchronousPluginMock);
+    });
+
+    it('should load the plugin commands', () => {
+      pluginManager.addPlugin(SynchronousPluginMock);
+
+      expect(pluginManager.commandsList[0]).to.have.property('deploy');
+    });
+  });
+
+  describe('#loadAllPlugins()', () => {
+    it('should load only core plugins when no service plugins are given', () => {
+      // Note: We need the HelloWorld plugin for this test to pass
+      pluginManager.loadAllPlugins();
+
+      expect(pluginManager.plugins).to.include(helloWorld);
+    });
+
+    it('should load all plugins when service plugins are given', () => {
```

Comment:
Could you add an additional test (or simply test it here) that the order of the plugins is properly preserved. Basically they should be run in the order they are added to the Plugin list and we have to test that this is the case